File: JavaScript/648-Replace Words.js
/**
 * @param {string[]} dictionary
 * @param {string} sentence
 * @return {string}
 */
class TrieNode{
    constructor(){
        this.children = new Map();
        this.isWord = false;
    }
}

var replaceWords = function(dictionary, sentence) {
    const root = new TrieNode();
    for (let word of dictionary){
        let node = root;
        for (let char of word){
            if (!node.children.has(char)){
                node.children.set(char, new TrieNode())
            }
            node = node.children.get(char);
        }
        node.isWord = true;
    }
    
    let search = (word, node)=>{
        let tmp = "";
        for (const char of word){
            if (node.children.has(char) && node.children.get(char).isWord){
                tmp += char;
                return tmp;
            }
            else if (node.children.has(char)){
                tmp += char;
                node = node.children.get(char);
            }
            else{
                return null;
            }
        }
        return null;
    }
    
    
    
    let res = sentence.split(" ");
    for (let i=0; i<res.length; i++){
        let replace = search(res[i], root);
        if (replace){
            res[i] = replace;
        }
    }
    
    return res.join(" ");
};

File: C++/648-Replace Words.cpp
struct TrieNode {
    TrieNode* ch[26] = {};  // Initialize all pointers to nullptr
    bool isWord = false;
};

class Solution {
public:
    string replaceWords(vector<string>& dictionary, string sentence) {
        TrieNode* t = new TrieNode();
        // build Trie
        for (string dict_word:dictionary) {
            TrieNode* node = t;
            for (char c:dict_word) {
                if (node->ch[c-'a'] == nullptr) {
                    node->ch[c-'a'] = new TrieNode();
                }
                node = node->ch[c-'a'];
            }
            node->isWord = true;
        }
        
        stringstream ss(sentence);
        string word;
        string res = "";
        while (ss >> word) {
            TrieNode* node = t;
            string curr = "";
            bool notFound = false;
            for (char w:word) {
                node = node->ch[w-'a'];
                if (node == nullptr) {
                    notFound = true;
                    break;
                }
                curr += w;
                if (node->isWord) break;
            }
            res += (notFound?word:curr) + " ";
        }
        res.pop_back();
        return res;
    }
};